### PR TITLE
feat: admission webhook sidecar rewrite

### PR DESCRIPTION
**Pull request by NohaIhab**
## Summary:
- define Pebble layer
- use KubernetesResourceHandler to manage k8s resources (crds and auth_manifests)
- use KubernetesServicePath library to patch the admission webhook service
- define on_event handler that performs all required actions for the charm
- rewrite cert generation to be more robust inspired by [canonical/namespace-node-affinity-operator](https://github.com/canonical/namespace-node-affinity-operator)
- upload certs to container with container.push
- define on_remove handler to remove resources safely
- add integration tests for on_remove
- create unit tests for: Not leader scenario, No relation scenario, Pebble layer, deploy K8s resources, generate certs

### TODO in another PR:
- implement health check on admission webhook service
- handle update status event to report health check status
- handle and test upgrade event

### CODE REVIEW

**Comment by i-chvets** [comment]
Will take a look when the work is complete.

File: src/charm.py [hunk 1]
```diff
@@ -1,303 +1,312 @@
...
+                self.unit.status = MaintenanceStatus("Force applying CRD resources")
+                self.logger.warning("Apply CRD resources with forced changes against conflicts")
+                self.crd_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("CRD resources creation failed") from error
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    def _gen_certs_if_missing(self):
+        """Generate certificates if they don't already exist in _stored."""
+        self.logger.info("_gen_certs_if_missing")
```
Comment (beliaev-maksim) [inline src/charm.py:200]: looks as not user friendly

File: src/charm.py [hunk 2]
```diff
@@ -1,303 +1,312 @@
...
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            raise CheckFailed(f"{e.status.message}", e.status_type)
-        return image_details
+            self._check_container_connection(self.container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
+
+        self.container.push(CONTAINER_CERTS_DEST + "/key.pem", self._cert_key, make_dirs=True)
```
Comment (beliaev-maksim) [inline src/charm.py:224]: use pathlib to generate paths?

File: src/charm.py [hunk 3]
```diff
@@ -1,303 +1,312 @@
...
+
+        self.unit.status = MaintenanceStatus("K8S resources removed")
+
+    def _on_pebble_ready(self, event):
+        """Configure started container."""
+        try:
+            self._check_container_connection(self.container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
```
Comment (beliaev-maksim) [inline src/charm.py:274]: you already do check inside upload certs
```suggestion
```

File: src/templates/auth_manifests.yaml.j2 [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+metadata:
+  name: {{ app_name }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: {{ app_name }}
+subjects:
+- kind: ServiceAccount
+  name: {{ app_name }}
+  namespace: {{ namespace }}
```
Comment (beliaev-maksim) [inline src/templates/auth_manifests.yaml.j2:31]:
```suggestion
  namespace: {{ namespace }}

```

File: src/templates/ssl.conf.j2 [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+DNS.3 = {{ service_name }}.{{ model }}.svc
+DNS.4 = {{ service_name }}.{{ model }}.svc.cluster
+DNS.5 = {{ service_name }}.{{ model }}.svc.cluster.local
+IP.1 = 127.0.0.1
+[ v3_ext ]
+authorityKeyIdentifier=keyid,issuer:always
+basicConstraints=CA:FALSE
+keyUsage=keyEncipherment,dataEncipherment,digitalSignature
+extendedKeyUsage=serverAuth,clientAuth
+subjectAltName=@alt_names
```
Comment (beliaev-maksim) [inline src/templates/ssl.conf.j2:28]:
```suggestion
subjectAltName=@alt_names

```

**Review by beliaev-maksim — CHANGES_REQUESTED** [review]
see comments

File: src/charm.py [hunk 6]
```diff
@@ -1,303 +1,311 @@
...
+                getattr(self._stored, cert_attribute)
+            except AttributeError:
+                self._gen_certs()
+                break
+        self.logger.info("Certificates already exist.")
+
+    def _gen_certs(self):
+        """Refresh the certificates, overwriting them if they already existed."""
+        self.logger.info("Generating certificates..")
+        certs = gen_certs(model=self._namespace, service_name=f"{self._service_name}")
```
Comment (beliaev-maksim) [inline src/charm.py:215]:
```suggestion
        certs = gen_certs(model=self._namespace, service_name=self._service_name)
```

isn't name by default string ?

File: src/charm.py [hunk 7]
```diff
@@ -1,303 +1,311 @@
...
+    def _upload_certs_to_container(self):
+        """Upload generated certs to container."""
+        try:
+            self._check_container_connection(self.container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
+
+        self.container.push(
+            Path(f"{CONTAINER_CERTS_DEST}/key.pem"), self._cert_key, make_dirs=True
```
Comment (beliaev-maksim) [inline src/charm.py:228]: I think it should be different

```suggestion
CONTAINER_CERTS_DEST = Path(...)
CONTAINER_CERTS_DEST / "key.pem", self._cert_key, make_dirs=True
```

File: src/certs.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        )
+
+        ret_certs = {
+            "cert": Path(tmp_dir + "/cert.pem").read_text(),
+            "key": Path(tmp_dir + "/server.key").read_text(),
+            "ca": Path(tmp_dir + "/ca.crt").read_text(),
+        }
+
+        # cleanup temporary files
+        check_call(["rm", "-f", tmp_dir + "/cert-gen-*"])
```
Comment (beliaev-maksim) [inline src/certs.py:84]: remove via pathlib?

File: src/charm.py [hunk 9]
```diff
@@ -1,303 +1,309 @@
...
+    @property
+    def _admission_webhook_layer(self) -> Layer:
+        """Create and return Pebble framework layer."""
+        layer_config = {
+            "summary": "admission-webhook layer",
+            "description": "Pebble config layer for admission-webhook-operator",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Entry point of admission-webhook-operator image",
```
Comment (beliaev-maksim) [inline src/charm.py:132]: I think it is not an Entry

that is the description of the service, no?

File: src/charm.py (hunk 3 above)
Comment (i-chvets) [reply src/charm.py]: Even though comment is correct, in context of Pebble function we don't know that `_upload_certs_to_container` checks for connection. The connection check here is more to protect eventual `_on_event` call at the end of `_on_pebble_ready` function.
Upload certs function can be re-used in other places, like upgrade.
This is how it was done in Seldon around this https://github.com/canonical/seldon-core-operator/blob/main/src/charm.py#L352
If there is a better solution let's discuss it.
Either way, LGTM.

File: src/charm.py (hunk 9 above)
Comment (i-chvets) [reply src/charm.py:132]: Yeah, it is being copied from some earlier sidecar charm and it is spread all over. Let's change it.

Comment (i-chvets) [reply src/charm.py:132]: May be at the top of the charm:
```
METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
```
And then we can use it:
```
            "services": {
                self._container_name: {
                    "override": "replace",
                    "summary": METADATA["summary"],
```

Comment (beliaev-maksim) [reply src/charm.py:132]: I think summary for the charm and for the service is different :)
at least in other charms it is done differently

File: src/charm.py (hunk 3 above)
Comment (beliaev-maksim) [reply src/charm.py]: on event does not require container

and if it does, then we need to put it into the function that needs it

File: src/charm.py (hunk 9 above)
Comment (i-chvets) [reply src/charm.py:132]: Yes. Agreed. There could be more than one container per charm. Need to come up with summary for this one.

**Review by beliaev-maksim — APPROVED** [review]

File: src/charm.py (hunk 3 above)
Comment (NohaIhab) [reply src/charm.py]: Making it a point to discuss when we need to check container connection.

**Pull request merged**